src/magic_folder/config.py

```diff
@@ -204,40 +225,102 @@ class MagicFolderConfig(object):
+                "version (wanted {}, got {})".format(
+                    self.name,
+                    _magicfolder_config_version,
+                    dbversion,
                 )
+            )
 
     @property
-    def author(self):
-        with self.database:
-            cursor = self.database.cursor()
-            cursor.execute("SELECT author_name, author_private_key FROM config");
-            name, keydata = cursor.fetchone()
-            return LocalAuthor(
-                name=name,
-                signing_key=SigningKey(keydata, encoder=Base32Encoder),
-            )
+    @with_cursor
+    def author(self, cursor):
+        cursor.execute("SELECT author_name, author_private_key FROM config");
+        name, keydata = cursor.fetchone()
+        return LocalAuthor(
+            name=name,
+            signing_key=SigningKey(keydata, encoder=Base32Encoder),
+        )
 
     @property
-    def stash_path(self):
-        with self.database:
-            cursor = self.database.cursor()
-            cursor.execute("SELECT stash_path FROM config");
-            path_raw = cursor.fetchone()[0]
-            return FilePath(path_raw)
+    @with_cursor
+    def stash_path(self, cursor):
+        cursor.execute("SELECT stash_path FROM config");
+        path_raw = cursor.fetchone()[0]
+        return FilePath(path_raw)
+
+    @with_cursor
+    def get_local_snapshot(self, cursor, name, author):
+        """
+        return an instance of LocalSnapshot corresponding to
+        the given name and author. Traversing the parents
+        would give the entire history of local snapshots.
+
+        :param unicode name: magicpath that represents the relative path of the file.
+
+        :param author: an instance of LocalAuthor
+
+        :returns: An instance of LocalSnapshot for the given magicpath.
+        """
+        cursor.execute("SELECT snapshot_blob FROM local_snapshots"
+                       " WHERE path=?",
+                       (name,))
+        row = cursor.fetchone()
+        if not row:
+            return None
+        else:
+            return LocalSnapshot.from_json(row[0], author)
+
+    @with_cursor
+    def store_local_snapshot(self, cursor, snapshot):
+        """
+        Store or update the given local snapshot.
+
+        :param LocalSnapshot snapshot: The snapshot to store.
+        """
+        # insert a new row or update an existing row with the new blob.
+        cursor.execute(
+            """
+            INSERT INTO
+                [local_snapshots] ([path], [snapshot_blob])
+            VALUES
+                (?, ?)
+            ON CONFLICT
+                ([path])
+            DO UPDATE SET
+                [snapshot_blob] = excluded.[snapshot_blob]
+
+            """,
+            (snapshot.name, snapshot.to_json()),
+        )
+
+    @with_cursor
+    def _get_all_relpaths_from(self, cursor, statement):
+        """
+        :param unicode tablename: table name to fetch the relpaths from.
```

Comment:
The docs and params don't seem to match the function name or what it does .. seems more like `_set_of_results_from_sql_statement` or something..? Since it seems to only be used in the method below I would just combine them (and could switch to a helper when there's more than one user).